metrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,296 @@
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prebid/prebid-cache/config"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	StatusKey    string = "status"
+	FormatKey    string = "format"
+	ConnErrorKey string = "connection_error"
+
+	ConnOpenedKey string = "connection_opened"
+	ConnClosedKey string = "connection_closed"
+
+	TotalsVal     string = "total"
+	ErrorVal      string = "error"
+	BadRequestVal string = "bad_request"
+	JsonVal       string = "json"
+	XmlVal        string = "xml"
+	DefinesTTLVal string = "defines_ttl"
+	InvFormatVal  string = "invalid_format"
+	CloseVal      string = "close"
+	AcceptVal     string = "accept"
+
+	MetricsPrometheus = "Prometheus"
+)
+
+type PrometheusMetrics struct {
+	Registry    *prometheus.Registry
+	Puts        *PrometheusRequestStatusMetric
+	Gets        *PrometheusRequestStatusMetric
+	PutsBackend *PrometheusRequestStatusMetricByFormat
+	GetsBackend *PrometheusRequestStatusMetric
+	Connections *PrometheusConnectionMetrics
+	ExtraTTL    *PrometheusExtraTTLMetrics
+	MetricsName string
+}
+
+type PrometheusRequestStatusMetric struct {
+	Duration      prometheus.Histogram
+	RequestStatus *prometheus.CounterVec
+}
+
+type PrometheusRequestStatusMetricByFormat struct {
+	Duration           prometheus.Histogram
+	PutBackendRequests *prometheus.CounterVec
+	RequestLength      prometheus.Histogram
+}
+
+type PrometheusConnectionMetrics struct {
+	ConnectionsErrors *prometheus.CounterVec
+	ConnectionsClosed prometheus.Counter
+	ConnectionsOpened prometheus.Counter
+}
+
+type PrometheusExtraTTLMetrics struct {
+	ExtraTTLSeconds prometheus.Histogram
+}
+
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	timeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	requestSizeBuckets := []float64{0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
```

Comment:
This needs to be updated because the current range is between 0 and 1 bytes. I'll assume the actual size ranges we're monitoring is close to what others hosts are seeing. I recommend:

0, 4kb, 8kb, 16kb, 32kb, 64kb, 128kb, 256kb, 512kb, 1mb

`[]float64{0, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 393216, 524288, 1048576}`